### DISCUSSION TITLE
How to do audio search？

### DISCUSSION THREAD

**Discussion by luliloveshe** [milvus-io/milvus#27709] [Q&A and General discussion]
I have an audio library. I want to implement the function of searching for similar audio resources through audio clips. what do I do?

**Comment by yhmo** [comment]
1. convert your audio clips to embeddings by PANNs(search "PANNs" on google for more info)
2. insert the embeddings into Milvus, each embedding has a unique ID, store the ids and audio storage path into MySQL.
3. when you want to search an audio, convert the audio by PANNs to an embedding, use the embedding to search in Milvus, the milvus returns the most similar items to you
4. use the id of the most similar items to query the MySQL, get the audio storage path

**Comment by luliloveshe** [comment]
Thank you so much. In addition, I found relevant examples on towhee and a list of models used to extract vectors. I am currently wondering which model is better to choose.
Example url https://github.com/towhee-io/examples/blob/main/audio/audio_classification/music_genre_classification.ipynb
Model url https://towhee.io/tasks/detail/operator?field_name=Audio&task_name=Audio-Embedding
![a1](https://github.com/milvus-io/milvus/assets/32316248/cbee068e-63a5-4fae-b175-e18f7c91fe19)
![a2](https://github.com/milvus-io/milvus/assets/32316248/914ddf67-4ac7-4896-921c-4cc9b97c7cc5)

**Reply by luliloveshe** [reply]
There are PANNs on towhee, which can be directly used to extract feature vectors. Confused about whether to use PANNs, VGGish or other models?
https://towhee.io/audio-classification/panns
![9C}94RJ}C5IYC%2VL428~R9](https://github.com/milvus-io/milvus/assets/32316248/24fba32a-c593-4afa-8a4c-0eb9bfa11fd9)

**Reply by zc277584121** [reply]
@luliloveshe hello,
-  panns is mainly used for audio classification. This classification ability is more suitable for finding music with similar themes. Please note that no matter how long the audio length is, the embedding output of this operator is 2048 dimensions.
- The output embedding dimension of vggish is related to the input audio length. The audio embedding operator converts an input audio into a dense vector which can be used to represent the audio clip's semantics. Each vector represents for an audio clip with a fixed length of around 0.9s. It is more suitable for representing the information of small clips, rather than the category information of the entire audio.

**Reply by luliloveshe** [reply]
@zc277584121
My audio library contains a large number of minutes of music and seconds of sound effects. Search through user-uploaded audio. According to your reply, I think panns is the most suitable. After all, it can extract vectors for the complete audio. What do you think?

**Reply by zc277584121** [reply]
@luliloveshe 
You need to clearly distinguish whether you are doing "similar audio category retrieval" or "audio fingerprint retrieval" in your business.
The former extracts an overall audio vector from an audio, and the information it contains is very rough.
The latter is to cut the entire audio into multiple small clips, and then extract a small vector for each small clip. For the entire audio, the information it contains is at a very fine granularity. It can be used for audio duplication checking or copyright protection.
Please check https://github.com/towhee-io/examples/tree/main/audio for more informations.

**Comment by luliloveshe** [comment]
@zc277584121 
Don't care about the category, just search for similar audio, which should be the latter "audio fingerprint retrieval" you mentioned. Which model is better to choose?

**Comment by luliloveshe** [comment]
I looked at the https://github.com/towhee-io/examples/tree/main/audio example, using NNFp and VGGish both returning a set of fragment vectors. My scenario is: there are 100,000+ audio libraries with varying durations. Users can upload any audio and search for similar audios from the audio library. Then my scenario requires extracting vectors from the entire audio, which is not limited by duration.
 The data2vec model seems to meet the requirements, extracting the entire audio and returning a fixed 768-dimensional vector. Can I use data2vec to extract vectors and save them to milvus for search?
![HL6~3M ~JE0}7W%FL2 YTZD](https://github.com/milvus-io/milvus/assets/32316248/be2accaa-5267-4f9f-a3c3-d01c2199dcc9)
![8 1%R)BHXSUALLYJ1 GQI B](https://github.com/milvus-io/milvus/assets/32316248/9b8a8fb2-fd14-4d4f-bf90-23fa9c7502e6)
![D_%XM UA){VQI%I{G}15Z}O](https://github.com/milvus-io/milvus/assets/32316248/01a7bfbc-dd84-428c-8ce8-efac7dbbe0cf)

**Reply by zc277584121** [reply]
@luliloveshe 
I think you should use audio fingerprint retrieval related models, such as vggish or nnfp.
There is no need to worry about inconsistent embedding sizes caused by different audio lengths. 
https://github.com/towhee-io/examples/blob/main/audio/audio_fingerprint/audio_fingerprint_advanced.ipynb
This notebook introduces the temporal_network technology, which can solve this problem to match audio of different lengths and finally locate similar fragments. You can follow this notebook to give it a try.

**Reply by luliloveshe** [reply]
@zc277584121 
Thanks a lot, I'll take a look at this sample code. In addition, do you think the data2vec model is feasible? As shown in the screenshot, even if you randomly specify an audio, it can return a 768-dimensional vector.
![SCTML~BGIIVIAZGA~(5BSEW](https://github.com/milvus-io/milvus/assets/32316248/63d22057-a63b-4b4a-8e8d-f7fc5932dfdd)

**Reply by zc277584121** [reply]
@luliloveshe 
I don't recommend using it if the audio lengths in your dataset vary greatly and you want to match the audio more accurately.
But if your audios are of similar length and you want to search for audios with similar styles, you can use it.

**Reply by luliloveshe** [reply]
@zc277584121 
The audio styles and durations in my audio library vary widely. There are various styles, ranging from a few seconds to a few minutes, tens of minutes or even hours.

**Reply by zc277584121** [reply]
@luliloveshe 
If a one-hour audio only generates a (768,)-dimensional embedding, it must have lost a lot of information, which is very detrimental to retrieval.
So in this scenario, I recommend you to use the way of fingprint.

**Comment by luliloveshe** [comment]
@zc277584121 
Ok thanks, let me take a look at the example you recommended first. In addition, after searching for results, such as 40 similar audios, how to remove duplicates?

**Reply by zc277584121** [reply]
It depends on your business logic

**Comment by luliloveshe** [comment]
@zc277584121 
I've seen the example you recommended, and the temporal_network technology shouldn't be used because I don't need to be precise about the start and end time of the overlapping part. The example "Audio Fingerprint I: Build a Demo with Towhee & Milvus" will suffice. This example first totals the scores by Path, and then sorts them in reverse order.

**Comment by luliloveshe** [comment]
@zc277584121 
https://github.com/towhee-io/examples/blob/main/audio/audio_fingerprint/audio_fingerprint_beginner.ipynb
I tested it using this example.
Extract dozens of audio vectors to milvus, and then use one of the audio A as the search audio. It is expected that the same audio A should be returned, but another audio B is returned. I analyzed the source code and the vote method is to accumulate scores. Audio duration A is short, and few segment vectors are extracted, so few similar segment vectors are found. Although the score of a single segment vector is high, the total score is low; audio duration B is very long, and many segment vectors are extracted, and similar segment vectors are found just in time. There are also many, although the individual fragment vector scores are low, the total score is high. So the final search result is B audio, which is obviously unreasonable. How to deal with this situation?
16s - 恢弘大气16秒_爱给网_aigei_com.mp3 is audio A
25min57s - 梁祝（完全版）_爱给网_aigei_com.mp3 is audio B
![N`0B43$4)@I QU440Y`3M13](https://github.com/milvus-io/milvus/assets/32316248/a36a42ad-9d38-4dc0-aa16-f27e2f003185)
